### ISSUE TITLE
Large numbers of emojis essentially halt the transcription

### ISSUE THREAD

**Issue by lukasthaler** [mahtoid/DiscordChatExporterPy#15]
**Issue description**
If there is a large quantity of emojis in the messages, the transctiption process exponentially slows down to the point where it takes several minutes to complete, while making the bot unresponsive to any other input for the duration

**Steps to reproduce**
Get yourself a channel and spam about 150-200 emojis (doesn't matter if in one or multiple messages, doesn't matter if they are all the same or different), then try to transcribe the messages containing these emojis. The emojis alone will cause the transcript to process for about 20 seconds

**Suspected cause**
I did some digging in the source code and suspect that ultimately, `valid_scr` from `emoji_convert.py` is the culprit. This function executes a *synchronous* HTTP HEAD request to twemoji.maxcdn.com, and from what I have been able to see in the rest of the library, it will do so for each and every individual emoji that is present in the channel

**Suggested fix**
The obvious fix would be an emoji cache to avoid duplicate calls to twemoji, thereby drastically reducing the number of HTTP calls. This alone should deal with 80% of the issue. Switching to asynchronous calls (e.g. using aiohttp instead of the requests module) should help to further improve the performance and ensure that the bot stays responsive throughout the transcription process (in the worst case of 200 different emojis, asynchronous HTTP calls will be necessary)

**Comment by mahtoid** [comment]
Firstly, beautiful issue request. I have to love it when I see these come through.

You're correct, it's currently synchronous which can cause a stack up and delay other tasks. As you suggested, switching to `aiohttp` will be beneficial so I will set some time this week in to switching that in to allow for asynchronous usage. 

Additionally, establishing an emoji cache to avoid duplicate calls sounds like a good solution, it may be a little overkill for this type of "plugin" and would be a nuisance to some of the smaller users on this module. However I like your thinking, and something like that would work at scale. 

Sorry for the delay in responding to your issue.

**Comment by mahtoid** [comment]
`aiohttp` won't be a solution. After some time thinking about it, even though you would be turning the function asynchronous - it'll still be working in a synchronous manner as you can only send one request of at a time as technically we don't know the next part of the message. 

In terms of caching, functools gives access to the `lru_cache` decorator which pretty much solves the issue. It knocks the wait time from ~10 seconds to subsecond. I will need to just test it a little more, as I know their `cache` decorator (which returns `lru_cache` is for Python 3.9>) - so I'll need to just make sure there is no drastic change coming through which would cause EOL to some people unless they update (I respect the fact some people version lock for production purposes).

**Comment by lukasthaler** [comment]
I agree that going asynchronous won't speed up the transcription itself, but it would make the bot responsive throughout the process. If LRU-caching is in fact powerful enough to push the performance into the sub-second range, that should not be a huge factor, however.
The notable exception is the presence of a large number of different emojis, as this cannot be mitigated by caching and **will** cause a lot of HTTP requests. For that reason, I'd still view `aiohttp` as a nice to have, though maybe no longer strictly necessary, but that's your call to make, not mine

**Comment by mahtoid** [comment]
That is true. I've got a patch written up, I'll push it to Github later today (pypi version will be later this week as I need to get out another minor before that). 

`lru_cache` from functools doesn't work on asynchronous functions - so I kindly borrowed a cache function from dpy and stripped it just what the module will need to avoid bloat. Additionally switched out `requests` for `aiohttp`.

**Comment by lukasthaler** [comment]
That sounds very promising, thanks for your effort! I'll upgrade to the Github version and verify that this deals with the issue at hand some time in these next few days if you want me to (and haven't already done all of the testing yourself)
